apps/src/blockly/addons/cdoUtils.js

```diff
@@ -14,27 +14,146 @@ import {
+    experiments.isEnabled(experiments.MODAL_FUNCTION_EDITOR)
+  ) {
+    procedureTypesToHide.push('procedures_defnoreturn');
+  }
+  moveHiddenProcedures(
+    parsedSource,
+    parsedHiddenDefinitions,
+    procedureTypesToHide
+  );
+  return {parsedSource, parsedHiddenDefinitions, blockOrderMap};
+}
+
+/**
+ * Convert source and hidden definitions to parsed json objects. If source was xml,
+ * convert to json before parsing. Also create a block order map if source was xml, which will
+ * allow us to correctly place blocks on the workspace.
+ * @param {string} source - workspace serialization, either XML or JSON
+ * @param {[string]} hiddenDefinitions - hidden definitions serialization, in JSON. Only used in Google Blockly labs.
+ * @returns {parsedSource: Object, parsedHiddenDefinitions: Object, blockOrderMap: Object}
+ */
+function parseSourceAndHiddenDefinitions(source, hiddenDefinitions) {
   let isXml = stringIsXml(source);
-  let stateToLoad;
+  let parsedSource;
+  let parsedHiddenDefinitions;
   let blockOrderMap;
+  hiddenDefinitions = hiddenDefinitions || '{}';
   if (isXml) {
     const xml = parseXmlElement(source);
-    stateToLoad = convertXmlToJson(xml);
+    parsedSource = convertXmlToJson(xml);
     blockOrderMap = Blockly.Xml.createBlockOrderMap(xml);
   } else {
-    stateToLoad = JSON.parse(source);
+    parsedSource = JSON.parse(source);
   }
-  Blockly.serialization.workspaces.load(stateToLoad, workspace);
-  positionBlocksOnWorkspace(workspace, blockOrderMap);
+  parsedHiddenDefinitions = JSON.parse(hiddenDefinitions);
+  return {parsedSource, parsedHiddenDefinitions, blockOrderMap};
+}
+
+/**
+ * Move hidden procedures from the source to the hidden definition object.
+ * These will be used to initialize the main and hidden definitions workspaces, respectively.
+ * Procedures are hidden if they have a type in the procedureTypesToHide array.
+ * In addition, copy the procedure model from the source
+ * object to the hidden definition object when moving a procedure.
+ * @param {Object} source Project source object, parsed from JSON.
+ * @param {Object} hiddenDefinitions Hidden Definition object, parsed from JSON (or an empty object)
+ * @param {Array<string>} procedureTypesToHide procedure types to move to procedures object.
+ * @returns void
+ */
+function moveHiddenProcedures(source, hiddenDefinitions, procedureTypesToHide) {
+  if (
+    procedureTypesToHide.length === 0 ||
+    !source.blocks ||
+    !source.blocks.blocks
+  ) {
+    return;
+  }
+  const blocksToHide = [];
+  const otherBlocks = [];
+  const sourceProcedures = source.procedures;
+  hiddenDefinitions.procedures ||= [];
+  source.blocks.blocks.forEach(block => {
+    if (procedureTypesToHide.includes(block.type)) {
+      blocksToHide.push(block);
+      // If we found a block to hide, also copy the procedure model
+      // for that block to the hidden definitions workspace.
+      const sourceProcedureModel = sourceProcedures.filter(
```

Comment:
nit: One (possibly clearer?) approach similar to what you did for the other function?
```
const correspondingProcedure = sourceProcedures.find(p => p.id === block.extraState.procedureId);
if (correspondingProcedure) {
  hiddenDefinitions.procedures.push(correspondingProcedure);
}
```
It's fine as-is, too, though.